core/market/decentralized/src/negotiation/requestor.rs

```diff
@@ -87,6 +93,58 @@ impl RequestorBroker {
         Ok(())
     }
 
+    pub async fn counter_proposal(
+        &self,
+        subscription_id: &SubscriptionId,
+        prev_proposal_id: &str,
+        proposal: &ClientProposal,
+    ) -> Result<String, ProposalError> {
+        // TODO: Everything should happen under transaction.
+        // TODO: Check if subscription is active
+        // TODO: Check if this proposal wasn't already countered.
+        let prev_proposal = self
+            .db
+            .as_dao::<ProposalDao>()
+            .get_proposal(prev_proposal_id)
+            .await
+            .map_err(|e| {
+                ProposalError::FailedGetProposal(prev_proposal_id.to_string(), e.to_string())
+            })?
+            .ok_or_else(|| {
+                ProposalError::ProposalNotFound(
+                    prev_proposal_id.to_string(),
+                    subscription_id.clone(),
+                )
+            })?;
+
+        if &prev_proposal.negotiation.subscription_id != subscription_id {
+            Err(ProposalError::ProposalNotFound(
+                prev_proposal_id.to_string(),
+                subscription_id.clone(),
+            ))?
+        }
+
+        let is_initial = prev_proposal.body.prev_proposal_id.is_none();
```

Comment:
I think we could move this request to PR: https://github.com/golemfactory/yagna/pull/412
Since I reuse some code between requestor and provider, it will be a little bit more complicated case there ;)